### PR TITLE
Fix pod probe sampling start time upon kubelet restarts

### LINKED ISSUE #118396: Pod readiness probe does not start when setting periodSeconds: 3600
### What happened?

When configuring the readiness probe for the container in the pod and setting the periodSeconds: 3600 the probe does not seem to run. 

          readinessProbe:
            exec:
              command:
                - bash
                - /opt/readyprobe.sh
            initialDelaySeconds: 90
            periodSeconds: 3600
            timeoutSeconds: 600
            failureThreshold: 2

It can take up to 40 minutes for the pod to report as ready.

Note: this pod is part of a stateful set.
Note: sometimes the system does behave as expected . however everytime a new node is provisioned, the issue starts happening again

### What did you expect to happen?

readyiness probe should start after 90 seconds evertime pod is scheduled

### How can we reproduce it (as minimally and precisely as possible)?

setup a stateful set with at least one pod and container running and readiness probe set as follows
          readinessProbe:
            exec:
              command:
                - bash
                - /opt/readyprobe.sh
            initialDelaySeconds: 90
            periodSeconds: 3600
            timeoutSeconds: 600
            failureThreshold: 2

### Anything else we need to know?

This is in aws eks. The node get allocated via aws autoscaler. setting up the pod triggers a scale up to provision a new node

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.9", GitCommit:"a1a87a0a2bcd605820920c6b0e618a8ab7d117d4", GitTreeState:"clean", BuildDate:"2023-04-12T12:16:51Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"windows/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.9-eks-0a21954", GitCommit:"eb82cd845d007ae98d215744675dcf7ff024a5a3", GitTreeState:"clean", BuildDate:"2023-04-15T00:37:59Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
</details>

### Cloud provider

<details>
AWS EKS VERSION 1.25
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by levikobi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
```release-note
1. As mentioned at #118396, pods with high periodSeconds duration at the readiness probe will not be ready
for a long time if scheduled on new provisioned nodes. Having a predefined max duration wait time for the
probing workers can solve the issue.
2. I've also noticed that when having an initialDelay > 0, the first probe call (not doProbe) will happen
only after periodSeconds has elapsed, instead of initialDelay. Meaning the timing of the run() function
is slightly off.
```

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/118396

#### Does this PR introduce a user-facing change?

```release-note
None
```
<!--

### CODE REVIEW

**Review by Tusenka — APPROVED** [review]

**Review by SataQiu — COMMENTED** [review]
/lgtm

File: pkg/kubelet/prober/worker.go [hunk 1]
```diff
@@ -153,10 +153,15 @@ func (w *worker) run() {
 	probeTickerPeriod := time.Duration(w.spec.PeriodSeconds) * time.Second
 
 	// If kubelet restarted the probes could be started in rapid succession.
-	// Let the worker wait for a random portion of tickerPeriod before probing.
+	// Let the worker wait for a random portion of tickerPeriod before probing,
+	// up to a maximum of 1 minute.
 	// Do it only if the kubelet has started recently.
 	if probeTickerPeriod > time.Since(w.probeManager.start) {
-		time.Sleep(time.Duration(rand.Float64() * float64(probeTickerPeriod)))
+		maxSleepDuration := time.Minute
```
Comment (aojea) [inline pkg/kubelet/prober/worker.go:160]: /hold

we should make these changes with a wider consensus, adding magic timeouts deep in the code is confusing.
There is also the other angle, this hitter mechanism is still more confusing, is it documented, is it tested?

we should get this better covered

Comment (levikobi) [reply pkg/kubelet/prober/worker.go:160]: I hear you @aojea.

1. I'll add some tests around the new functionality.
2. Do you want me to add some small section at https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes, explaining about possible probing delay when kubelet restarts?
3. I'll get rid of the sleep command, in the favor of a select statement, similar to the one used in probeLoop.

Please do let me know if you have other suggestions

**Comment by levikobi** [comment]
I've updated my commit, as well as the description to better describe the issue.
Reviews are welcome.

**Review by gjkim42 — COMMENTED** [review]
I think limiting the delay duration for starting the probe workers after the kubelet restart makes sense. Anything else should be removed from this PR.

/cc @matthyx

File: pkg/kubelet/prober/worker.go [hunk 2]
```diff
@@ -176,15 +166,50 @@ func (w *worker) run() {
...
+		kubeletRestartJitter := time.Duration(rand.Float64() * float64(maxSleepDuration))
+		if stopped := waitUntilStopTriggerOrTimeout(time.After(kubeletRestartJitter)); stopped {
+			return
+		}
+	}
+
+	// Execute the probe once, even before InitialDelaySeconds have passed. This is done in order
+	// to initialize the results manager with the initial value of the probe.
+	beforeProbeExecution := time.Now()
+	w.doProbe(ctx)
```
[the remark this answers is not in the document]
Comment (levikobi) [reply pkg/kubelet/prober/worker.go:198]: Regarding the misleading comment, I agree. I will change it.

Regarding the second point. In the current implementation the first doProbe call is also being called before initialDelaySeconds has passed, so the results count should be equal as before

File: pkg/kubelet/prober/worker.go [hunk 3]
```diff
@@ -176,15 +166,50 @@ func (w *worker) run() {
...
+
+	// Wait for the remaining time of InitialDelaySeconds.
+	initialDelay := time.Duration(w.spec.InitialDelaySeconds)*time.Second - time.Since(beforeProbeExecution)
+	if stopped := waitUntilStopTriggerOrTimeout(time.After(initialDelay)); stopped {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (gjkim42) [reply pkg/kubelet/prober/worker.go]: I understand, multiple issues are present here.
Even so, this looks not correct, as the initial delay seconds is about the number of seconds after the container has started.

[the remark this answers is not in the document]
Comment (gjkim42) [reply pkg/kubelet/prober/worker.go]: IMHO, you may separate this issue into multiple small issues, and fix each of them, as some of them need more consideration.
1. limit the start delay due to the kubelet start - looks easy
2. small `initialDelaySeconds` and big `periodSeconds` - may need refactoring
3. ...

**Pull request closed without merging**